osu.Game/Utils/ColourUtils.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Utils;
+using osuTK.Graphics;
+
+namespace osu.Game.Utils
+{
+    public static class ColourUtils
+    {
+        /// <summary>
+        /// Samples from a given linear gradient at a certain specified point.
+        /// </summary>
+        /// <param name="gradient">The gradient, defining the colour stops and their positions (in [0-1] range) in the gradient.</param>
+        /// <param name="point">The point to sample the colour at.</param>
+        /// <returns>A <see cref="Color4"/> sampled from the linear gradient.</returns>
+        public static Color4 SampleFromLinearGradient(IReadOnlyList<(float position, Color4 colour)> gradient, float point)
+        {
+            if (point < gradient[0].position)
+                return gradient[0].colour;
+
+            for (int i = 0; i < gradient.Count - 1; i++)
+            {
+                var startStop = gradient[i];
+                var endStop = gradient[i + 1];
+
+                if (point >= endStop.position)
+                    continue;
+
+                return Interpolation.ValueAt(point, startStop.colour, endStop.colour, startStop.position, endStop.position);
```

Comment:
This interpolation is taking place in sRGB space and not linear space - in other words, it is not gamma-corrected. [Web compensates for this](https://github.com/ppy/osu-web/blob/71fbab8936d79a7929d13854f5e854b4f383b236/resources/assets/lib/utils/beatmap-helper.ts#L26). This means that the colour values are potentially slightly off.

I checked colours for [this mapset](https://osu.ppy.sh/beatmapsets/1130581#osu/2396095) against web and got the following:

![2021-08-03-235142_268x386_scrot](https://user-images.githubusercontent.com/20418176/128091532-3a9cb514-f7b5-4c29-ab15-642d19059cfe.png)

`(lazer before)` colours are this PR. `(lazer after)` colours are this PR after performing gamma correction by interpolating in linear space:

```diff
diff --git a/osu.Game/Utils/ColourUtils.cs b/osu.Game/Utils/ColourUtils.cs
index 515963971d..b42e10cecb 100644
--- a/osu.Game/Utils/ColourUtils.cs
+++ b/osu.Game/Utils/ColourUtils.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System.Collections.Generic;
+using osu.Framework.Extensions.Color4Extensions;
 using osu.Framework.Utils;
 using osuTK.Graphics;
 
@@ -28,7 +29,10 @@ public static Color4 SampleFromLinearGradient(IReadOnlyList<(float position, Col
                 if (point >= endStop.position)
                     continue;
 
-                return Interpolation.ValueAt(point, startStop.colour, endStop.colour, startStop.position, endStop.position);
+                var startLinear = startStop.colour.ToLinear();
+                var endLinear = endStop.colour.ToLinear();
+
+                return Interpolation.ValueAt(point, startLinear, endLinear, startStop.position, endStop.position).ToSRGB();
             }
 
             return gradient[^1].colour;
```

While not perfect still, the `(after)` numbers are a lot closer (discrepancies of at most 1-2 for each colour component).

Possibly `Interpolation.ValueAt()` for colours should be doing the linear conversion automatically, in framework. No strong opinions here.

Reply:
Yeah... I'm interested to hear the other's opinions on having gamma correction in `Interpolation.ValueAt(Color4)` itself.

Reply:
I'm confused. How is interpolation done in the SRGB space? That would mean the input colours are coming in as SRGB.
As it is, doesn't this mean the following process is taking place?:
```
input colour (linear)
SampleFromLinearGradient() -> (assumes SRGB) .ToLinear() -> return colour
Drawable.Colour = value -> implicit conversion from Color4 to SRGColour -> (assumes SRGB) .ToLinear()
```
So there are two from-SRGB conversions taking place after the above patch, and one from-SRGB conversion taking place in the original changes.
Isn't the issue instead that this function doesn't output an SRGB colour? That is, changing to `return Interpolation.ValueAt(...).ToSRGB()`.

Reply:
I'll admit that I didn't 100% confirm what was going on, just wrote what would make sense to me and compared output, but if the source hex colours were taken from figma, then those would be srgb, right? Hence the first two conversions.

Reply:
In that case then I'd expect the `.ToLinear()` to be done locally where those colours are declared.

Reply:
Are you getting the values from the figma parameter display, or using an eye-dropper? I think these may give different values. I would presume the parameters listed would give the original values, right? We should never actually be using SRGB post conversion values.

Reply:
I have always been using the figma parameter display, and now after reading a bit on the sRGB/linear (hard-to-grasp) logic, the interpolation should be done in the linear space? unless the colours provided by figma are in the sRGB space...?

Reply:
@arflyte can you confirm whether the figma colours are sRGB? I’ve thought they were in the standard RGB colour space until now.

Reply:
Relevant docs: https://help.figma.com/hc/en-us/articles/360039825114-Color-management-in-the-desktop-app

> In the browser, Figma uses a managed sRGB color profile.

This is likely because the web at large also uses sRGB, so for the colours to match it only makes sense to do that for the CSS rules.

Reply:
> @arflyte can you confirm whether the figma colours are sRGB? I’ve thought they were in the standard RGB colour space until now.

I already went through this with flyte since he used to have his figma and photoshop set wrong. You can only assume all colours are in sRGB for all provided designs or nothing will make sense.

The code above by @bdach looks to be correct in what we want here.

Reply:
Right, one remaining question I have is whether we can have the proposed logic in `Interpolation.ValueAt(Color4)` itself?

Reply:
For now that is the safe and simple assumption if we are aiming to match web colours, yes.